backend/models/distance.py

```diff
@@ -66,11 +77,105 @@ def get_distance_in_miles(self, point_a, point_b):
+        return '{lng},{lat}'.format(lat=point['latitude'], lng=point['longitude'])
+
+    @retry(retry_on_result=_retry_if_result_none, stop_max_attempt_number=10, wait_fixed=2000)
+    def get_distance_in_miles(self, point_a, point_b):
+        """Use an OSRM server to compute the distance and time between two points."""
+        points_as_str = [self._represent_point_as_str(p) for p in (point_a, point_b)]
+        coordinates = ';'.join(points_as_str)
+        url = self.api_url + '/table/v1/car/{coords}'.format(coords=coordinates)
+        response = requests.get(url)
+
+        response.raise_for_status()
+        content = response.json()
+
+        if content['code'] == 'Ok':
+            if content['durations']:
+                return content['durations'][0][1] * ONE_METER_IN_MILES
+        else:
+            return None
+
+    def closest(self, origin, point_list):
+        """Find closest point in a list of points and returns min_distance, min_point."""
+        time.sleep(0.1)
+
+        origin_str = self._represent_point_as_str(origin)
+        coordinates = ';'.join(
+            [origin_str] + [self._represent_point_as_str(point) for point in point_list]
+        )
+
+        url = self.api_url + '/table/v1/car/{coords}'.format(coords=coordinates)
+        params = {'sources': 0}
+        response = requests.get(url, params=params)
+
+        response.raise_for_status()
+        content = response.json()
+
+        if content['code'] == 'Ok':
+            if content['durations']:
+                distance_responses = content['durations'][0][1:]
+        else:
+            return None
+
+        min_idx, min_distance = min(
+            enumerate(distance_responses),
+            key=operator.itemgetter(1)
+        )
+        # TODO: Determine the origin of these magic numbers to convert seconds to meters.
+        return min_distance * ONE_METER_IN_MILES * 200.0 / 11.1, point_list[min_idx]
+
+    def closest_with_early_exit(self, origin, point_list, exit_distance):
+        """
+        Find closest point in a list of points, exiting early if exit_distance is reached.
+
+        The exit_distance uses haversine distance.
+        """
+        haversine_distances = [
+            self._haversine_measurer.get_distance_in_miles(origin, point)
+            for point in point_list
+        ]
+        min_haversine_idx, min_haversine_distance = min(
+            enumerate(haversine_distances),
+            key=operator.itemgetter(1)
+        )
+
+        if min_haversine_distance <= exit_distance:
+            return min_haversine_distance, point_list[min_haversine_idx]
+        else:
+            relevant_points = [
+                point_list[idx]
+                for idx, distance in enumerate(haversine_distances)
+                if distance < min(self.early_exit_outer_radius, 1.5 * min_haversine_distance)
+            ]
+            if not relevant_points:
+                return 99999.9, point_list[0]
```

Comment:
Could we return infinity? Open question - how would we handle that in the frontend?